### ISSUE TITLE
libgit2 workdir to index diff reporting renamed submodules out of nowhere

### ISSUE THREAD

**Issue by swisspol** [libgit2/libgit2#3713]
I have a repo with 6 submodules set up the exact same way. With the latest libgit2, the diff between workdir and index (which is clear) reports that 3 of them have been "renamed" with an extra `/` at the end of their path:

![image](https://cloud.githubusercontent.com/assets/627915/14094603/bcbbd604-f50c-11e5-9c9d-b2a75818cbf4.png)

This is observed on OS X 10.11.4 using my app [GitUp](http://gitup.co) in which I just updated libgit2 from e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2 to e89abec69f114a4480431c74160558b86148af41.

@ethomson The major difference I immediately see is the major iterator refactor you did. Could this be related?

`git_diff_options` passed in to `git_diff_index_to_workdir()`:
```
(lldb) p *diffOptions
(git_diff_options) $1 = {
  version = 1
  flags = 33587224
  ignore_submodules = GIT_SUBMODULE_IGNORE_NONE
  pathspec = {
    strings = 0x0000000000000000
    count = 0
  }
  notify_cb = 0x0000000000000000
  progress_cb = 0x0000000000000000
  payload = 0x0000000000000000
  context_lines = 3
  interhunk_lines = 0
  id_abbrev = 0
  max_size = 8388608
  old_prefix = 0x0000000000000000 <no value available>
  new_prefix = 0x0000000000000000 <no value available>
}
```

`git_diff_delta` contents for `External/libPusher`:
```
(lldb) p *delta
(git_diff_delta) $1 = {
  status = GIT_DELTA_DELETED
  flags = 0
  similarity = 0
  nfiles = 1
  old_file = {
    id = (id = unsigned char [20] @ 0x00007fba9ffe5050)
    path = 0x00006210001d3960 "External/libPusher"
    size = 1088
    flags = 12
    mode = 57344
    ctime = 1442841017
    mtime = 1442841017
  }
  new_file = {
    id = (id = unsigned char [20] @ 0x00007fba9ffe5090)
    path = 0x00006210001d3960 "External/libPusher"
    size = 0
    flags = 4
    mode = 0
    ctime = 0
    mtime = 0
  }
}

(lldb) p *delta
(git_diff_delta) $2 = {
  status = GIT_DELTA_UNTRACKED
  flags = 0
  similarity = 0
  nfiles = 1
  old_file = {
    id = (id = unsigned char [20] @ 0x00007fbababcebb0)
    path = 0x00006210001d3978 "External/libPusher/"
    size = 0
    flags = 4
    mode = 0
    ctime = 0
    mtime = 0
  }
  new_file = {
    id = (id = unsigned char [20] @ 0x00007fbababcebf0)
    path = 0x00006210001d3978 "External/libPusher/"
    size = 1088
    flags = 8
    mode = 16384
    ctime = 1455907346
    mtime = 1455907346
  }
}
```

`git submodule` output:
```
 981347b6e432007882fd59f9a498c363a3b39ec0 External/Catch (v1.3.1)
 faec648b2547f99b909676d04e0f736a79593455 External/MTBBarcodeScanner (1.8.11)
 954c9478b5710903a485c68fe046e65293e2c900 External/SocketRocket (v0.2.0-158-g954c947)
 f64d469fa80d8203598924ee35bad88b83261959 External/libPusher (v1.6.1)
 9515b9cc949e19f6aa0b904aac54fb65661b3c6d External/rapidjson (v1.0.1-248-g9515b9c)
 f6729c6b3654431935b59aca859e15648b3f3ee1 External/raven-objc (0.3.0-65-gf6729c6)
```

`.git/config` content:
```
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true
[branch "master"]
	remote = origin
	merge = refs/heads/master
[remote "origin"]
	url = <REDACTED>
	fetch = +refs/heads/*:refs/remotes/origin/*
[submodule "External/SocketRocket"]
	url = https://github.com/square/SocketRocket.git
[submodule "External/libPusher"]
	url = https://github.com/lukeredpath/libPusher.git
[submodule "External/Catch"]
	url = https://github.com/philsquared/Catch.git
[submodule "External/rapidjson"]
	url = https://github.com/miloyip/rapidjson.git
[submodule "External/MTBBarcodeScanner"]
	url = https://github.com/mikebuss/MTBBarcodeScanner.git
[submodule "External/raven-objc"]
	url = https://github.com/getsentry/raven-objc.git
```

`.gitmodules` content:
```
[submodule "External/SocketRocket"]
	path = External/SocketRocket
	url = https://github.com/square/SocketRocket.git
[submodule "External/libPusher"]
	path = External/libPusher
	url = https://github.com/lukeredpath/libPusher.git
[submodule "External/Catch"]
	path = External/Catch
	url = https://github.com/philsquared/Catch.git
[submodule "External/rapidjson"]
	path = External/rapidjson
	url = https://github.com/miloyip/rapidjson.git
[submodule "External/MTBBarcodeScanner"]
	path = External/MTBBarcodeScanner
	url = https://github.com/mikebuss/MTBBarcodeScanner.git
[submodule "External/raven-objc"]
	path = External/raven-objc
	url = https://github.com/getsentry/raven-objc.git
```

Modules content:
```
$ ls -l .git/modules/External/
total 0
drwxr-xr-x+ 15 pol  staff  510 Mar 28 17:38 Catch
drwxr-xr-x+ 16 pol  staff  544 Mar 28 17:38 MTBBarcodeScanner
drwxr-xr-x+ 17 pol  staff  578 Mar 28 17:38 SocketRocket
drwxr-xr-x+ 15 pol  staff  510 Mar 28 17:38 libPusher
drwxr-xr-x+ 16 pol  staff  544 Mar 28 17:38 rapidjson
drwxr-xr-x+ 15 pol  staff  510 Mar 28 17:38 raven-objc
```

Checkout content:
```
$ ls -l External/
total 3296
drwxr-xr-x+  4 pol  staff      136 Mar  8 07:36 Android-Libraries
drwxr-xr-x+ 14 pol  staff      476 Feb 19 10:42 Catch
drwxr-xr-x+ 15 pol  staff      510 Feb 19 10:42 MTBBarcodeScanner
drwxr-xr-x+  5 pol  staff      170 Feb  9 15:48 OSX-Frameworks
drwxr-xr-x+ 22 pol  staff      748 Mar 25 13:53 SocketRocket
drwxr-xr-x+  5 pol  staff      170 Mar 26 19:05 iOS-Frameworks
drwxr-xr-x+ 32 pol  staff     1088 Feb 19 10:42 libPusher
drwxr-xr-x+  5 pol  staff      170 Feb 19 10:42 libexif
-rw-r--r--+  1 pol  staff  1686528 Feb 19 10:42 nuget.exe
drwxr-xr-x+ 25 pol  staff      850 Feb 19 10:42 rapidjson
drwxr-xr-x+ 13 pol  staff      442 Feb 23 12:05 raven-objc
drwxr-xr-x@  6 pol  staff      204 Feb 19 10:42 sqlite3
```

**Comment by swisspol** [comment]
FWIW I tried deleting `External/SocketRocket` (one of the "renamed" submodules) from `.git/config`, `.git/modules` and the checkout, then ran `git submodule update --init`, but the submodule is still showing as "renamed".

Then I tried the same operation with a non-renamed submodule (`External/Catch`), and it didn't change anything either.

I really don't understand where libgit2 is getting these trailing `/`.

**Comment by swisspol** [comment]
Stepping through `git_diff__from_iterators()` until it goes through `External/libPusher`, this is what we have:
```
(lldb) p *info.oitem
(git_index_entry) $2450 = {
  ctime = (seconds = 1442841017, nanoseconds = 0)
  mtime = (seconds = 1442841017, nanoseconds = 0)
  dev = 0
  ino = 96743676
  mode = 57344
  uid = 502
  gid = 20
  file_size = 1088
  id = (id = unsigned char [20] @ 0x00007fbabb0af6b8)
  flags = 18
  flags_extended = 0
  path = 0x000060b000123470 "External/libPusher"
}
(lldb) p *info.nitem
(git_index_entry) $2451 = {
  ctime = (seconds = 1455907346, nanoseconds = 0)
  mtime = (seconds = 1455907346, nanoseconds = 0)
  dev = 16777220
  ino = 96743676
  mode = 16384
  uid = 502
  gid = 20
  file_size = 1088
  id = (id = unsigned char [20] @ 0x00007fba87c68a28)
  flags = 0
  flags_extended = 0
  path = 0x00006210002a214c "External/libPusher/"
}
```

And the `cmp` variable is `-47`.

The question here is why is the workdir iterator returning a trailing `/` for some submodules on disk and not others?

**Comment by swisspol** [comment]
Now looking into `filesystem_iterator_frame_push()` when it's calling `filesystem_iterator_is_submodule()` on the path `External/libPusher`, it appears that `git_index_snapshot_find()` returns `GIT_ENOTFOUND`, which means libgit2 thinks `External/libPusher` is not a submodule but a regular directory.

**Comment by swisspol** [comment]
I tried deleting `.git/index` and having Git recreate it by doing `git reset --hard`, but the bug is still here, so I don't think the problem is in the Git index.

**Comment by swisspol** [comment]
Submodule entry in index that libgit2 does not identify correctly:
```
$ git ls-files --stage External/libPusher
160000 f64d469fa80d8203598924ee35bad88b83261959 0	External/libPusher
```

Submodule entry in index that libgit2 identifies correctly:
```
$ git ls-files --stage External/rapidjson
160000 9515b9cc949e19f6aa0b904aac54fb65661b3c6d 0	External/rapidjson
```

I don't know what else to do here, any help would be much appreciated!

**Comment by swisspol** [comment]
Interestingly a clean checkout of that repo followed by `git submodule update --init` reproduces the bug.

Outside this private repo, the [repo of GitUp](https://github.com/git-up/GitUp) itself seems affected if you do a clean checkout: 

![image](https://cloud.githubusercontent.com/assets/627915/14095706/d75559ee-f517-11e5-8a4b-8603b8e7e536.png)

```
$ git submodule
 04686418418ca90581233690f8edc320c5975aa1 GitUpKit/Third-Party/Libraries (heads/gitup)
 06607e551328c453191a179f980818c2ea888932 GitUpKit/Third-Party/XLFacility (1.5-28-g06607e5)
 422482d88312f0c88390ba6b4920b9bf6d639189 GitUpKit/Third-Party/libgit2 (v0.16.0-7392-g422482d)

$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
```

That's good news because this means there's a publicly reproducible case!

**Comment by johnhaley81** [comment]
@swisspol is this fixed by #3583?

**Comment by swisspol** [comment]
I don't see why it would be considering this PR only touches the checkout code which is not the problem here.

**Comment by swisspol** [comment]
@ethomson Do you have any clue about what is going on or what else I could do to figure out why some submodules are affected and not others?

**Comment by carlosmn** [comment]
A screenshot from whatever application that is doesn't tell us what libgit2 is returning. If you could provide a test case, that would go a long way towards getting to what is happening in the library.

**Comment by ethomson** [comment]
I'll take a look at this today or tomorrow, but as @carlosmn mentioned, having more data closer to the library is always helpful.  Thanks!

**Comment by swisspol** [comment]
There:

```
$ git --version
git version 2.6.4 (Apple Git-63)
$ git clone --recursive https://github.com/git-up/GitUp.git
(...)
```

```
#include <git2.h>
#include <stdio.h>
#include <assert.h>

int main(int argc, const char * argv[]) {
  assert(git_libgit2_init() >= 1);
  git_repository* repository;
  git_index* index;
  git_diff* diff;
  int result;

  result = git_repository_open(&repository, "/Users/pol/Downloads/GitUp");  // git clone --recursive https://github.com/git-up/GitUp.git
  assert(result == GIT_OK);
  result = git_repository_index(&index, repository);
  assert(result == GIT_OK);
  git_diff_options options = GIT_DIFF_OPTIONS_INIT;
  options.flags |= GIT_DIFF_INCLUDE_UNTRACKED | GIT_DIFF_RECURSE_UNTRACKED_DIRS | GIT_DIFF_SHOW_UNTRACKED_CONTENT | GIT_DIFF_UPDATE_INDEX;
  options.ignore_submodules = GIT_SUBMODULE_IGNORE_NONE;
  result = git_diff_index_to_workdir(&diff, repository, index, &options);
  for (size_t i = 0; i < git_diff_num_deltas(diff); ++i) {
    const git_diff_delta* delta = git_diff_get_delta(diff, i);
    printf("%s | %s | %i\n", delta->old_file.path, delta->new_file.path, delta->status);
  }
  assert(result == GIT_OK);
  return 0;
}
```

Output:
```
GitUpKit/Third-Party/Libraries | GitUpKit/Third-Party/Libraries | 2
GitUpKit/Third-Party/Libraries/ | GitUpKit/Third-Party/Libraries/ | 7
GitUpKit/Third-Party/XLFacility | GitUpKit/Third-Party/XLFacility | 2
GitUpKit/Third-Party/XLFacility/ | GitUpKit/Third-Party/XLFacility/ | 7
Program ended with exit code: 0
```

vs

```
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
```

**Comment by swisspol** [comment]
@ethomson Thanks for fixing this rapidly. In the future, please add a comment to the issue, because, unless I missed it, I didn't get an email notification about the mention of #3719 in this topic, so I spent an hour identifying and building a test case for no reason, _after_ you did your PR :disappointed:

**Comment by carlosmn** [comment]
Fixed via #3719
